File: app/src/main/java/com/rakangsoftware/recyclerviewdemo/CountryAdapter.java
package com.rakangsoftware.recyclerviewdemo;

import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by pererikbergman on 2017-12-14.
 */

public class CountryAdapter extends RecyclerView.Adapter<CountryViewHolder> {

    private List<Country> mCountryList = new ArrayList<>();

    private OnCountryClickListener mOnCountryClickListener;

    public void setCountryList(final List<Country> countryList) {
        mCountryList = countryList;
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return mCountryList.size();
    }

    @Override
    public CountryViewHolder onCreateViewHolder(final ViewGroup parent, final int viewType) {
        return CountryViewHolder.newInstance(parent, mOnCountryClickListener);
    }

    @Override
    public void onBindViewHolder(final CountryViewHolder holder, final int position) {
        holder.bind(mCountryList.get(position));
    }


    public void setOnCountryClickListener(final OnCountryClickListener onCountryClickListener) {
        mOnCountryClickListener = onCountryClickListener;
    }
}

File: app/src/main/java/com/rakangsoftware/recyclerviewdemo/CountryAdapter.kt
package com.rakangsoftware.recyclerviewdemo

import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import java.util.*

/**
 * Created by pererikbergman on 2017-12-14.
 */

class CountryAdapter : RecyclerView.Adapter<CountryViewHolder>() {

    private var mCountryList: List<Country> = ArrayList()

    private var mOnCountryClickListener: OnCountryClickListener? = null

    fun setCountryList(countryList: List<Country>) {
        mCountryList = countryList
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int = mCountryList.size

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CountryViewHolder = CountryViewHolder.newInstance(parent, mOnCountryClickListener!!)

    override fun onBindViewHolder(holder: CountryViewHolder, position: Int) = holder.bind(mCountryList[position])

    fun setOnCountryClickListener(onCountryClickListener: OnCountryClickListener) {
        mOnCountryClickListener = onCountryClickListener
    }
}
